[NVPTX] Improve lowering of v4i8

Artem-B opened the pull request:
Make it a legal type and plumb through lowering of relevant instructions.

Artem-B committed:
[NVPTX] Improve lowering of v4i8

Make it a legal type and plumb through lowering of relevant instructions.

ThomasRaoux reviewed (approved):
LGTM

ThomasRaoux commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3292,6 +3293,10 @@ let hasSideEffects = false in {
                              (ins Int16Regs:$s1, Int16Regs:$s2,
                                   Int16Regs:$s3, Int16Regs:$s4),
                              "mov.b64 \t$d, {{$s1, $s2, $s3, $s4}};", []>;
+  def V4I8toI32  : NVPTXInst<(outs Int32Regs:$d),
+                             (ins Int16Regs:$s1, Int16Regs:$s2,
+                                  Int16Regs:$s3, Int16Regs:$s4),
+                             "mov.b32 \t$d, {{$s1, $s2, $s3, $s4}};", []>;
```
this generates ptx like this:
```
mov.b32         %r10, {%rs2, %rs1, %rs4, %rs3};
```

which cause ptxas to emit this error:
```
error   : Arguments mismatch for instruction 'mov'
```

ThomasRaoux replied:
is it because we use 16bits register to store 8 bits values?

Artem-B replied:
According to the docs .B3 2nd is expected to be able to pack/unpack four 8-bit values: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-mov-2

![image](https://github.com/llvm/llvm-project/assets/526795/0497ae25-a8c8-4395-a221-b1a09fb3d404)

However, they do expect the arguments to be a .b8 values. 

move.b32 to/from elements of `v4.b8` appears to work: https://cuda.godbolt.org/z/qros4r6r1

Artem-B replied:
Under the hood byte extraction/insertion ends up as BFI/BFE instructions, so we may as well just do that in PTX, too.

Artem-B replied:
Latest commint on this PR should no longer produce uncompileable PTX.

ThomasRaoux reviewed (changes requested):
Actually trying out the patch in triton it causes some invalid ptx to get emitted (see comment)

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -3307,6 +3312,10 @@ let hasSideEffects = false in {
+  def I32toV4I8  : NVPTXInst<(outs Int16Regs:$d1, Int16Regs:$d2,
+                                   Int16Regs:$d3, Int16Regs:$d4),
+                             (ins Int32Regs:$s),
+                             "mov.b32 \t{{$d1, $d2, $d3, $d4}}, $s;", []>;
```
Now I wonder if this works. Will check.

Artem-B committed:
More work on fleshing out extractelt/build_vector for v4i8

Verified that NVPTX tests pass with ptxas being able to compiler PTX produced by
llc tests.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -1781,6 +1794,24 @@ def : Pat<(i16 (sext_inreg (trunc Int64Regs:$s), i8)),
+         (CVT_s16_s32 (BFE_S32rii Int32Regs:$src, imm:$bitidx, 8), CvtNONE)>;
+
+
+def : Pat<(v4i8 (build_vector (i16 Int16Regs:$a), (i16 Int16Regs:$b),
+                              (i16 Int16Regs:$c), (i16 Int16Regs:$d))),
+          (BFI_B32rii
+           (BFI_B32rii (CVT_u32_u16 Int16Regs:$d, CvtNONE), (CVT_u32_u16 Int16Regs:$c, CvtNONE), 8, 8),
+           (BFI_B32rii (CVT_u32_u16 Int16Regs:$b, CvtNONE), (CVT_u32_u16 Int16Regs:$a, CvtNONE), 8, 8),
+           16, 16)>;
+
```
This still needs tests.

ThomasRaoux commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -491,6 +497,10 @@ NVPTXTargetLowering::NVPTXTargetLowering(const NVPTXTargetMachine &TM,
+  setOperationAction(ISD::VECTOR_SHUFFLE, MVT::v4i8, Expand);
+  setOperationAction(ISD::BUILD_VECTOR, MVT::v4i8, Custom);
+
   // Operations not directly supported by NVPTX.
```
It looks like we are missing some actions for v4i8 types like from line 639 to line 658.

```
  setI16x2OperationAction(ISD::ABS, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::SMIN, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::SMAX, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::UMIN, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::UMAX, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::CTPOP, MVT::v2i16, Legal, Expand);
  setI16x2OperationAction(ISD::CTLZ, MVT::v2i16, Legal, Expand);
  setI16x2OperationAction(ISD::ADD, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::SUB, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::MUL, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::SHL, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::SREM, MVT::v2i16, Legal, Custom);
  setI16x2OperationAction(ISD::UREM, MVT::v2i16, Legal, Custom);
  // Other arithmetic and logic ops are unsupported.
  setOperationAction({ISD::SDIV, ISD::UDIV, ISD::SRA, ISD::SRL, ISD::MULHS,
                      ISD::MULHU, ISD::FP_TO_SINT, ISD::FP_TO_UINT,
                      ISD::SINT_TO_FP, ISD::UINT_TO_FP},
                     MVT::v2i16, Expand);
```

Artem-B committed:
Down the rabbit hole we go.

To make things work consisstently for v4i8, we need to implement other vector
ops.

Artem-B committed:
Added vector_shuffle lowering to PRMT.

Artem-B committed:
Address clang-format complaints.

Artem-B committed:
Use .lo/ls/hi/hs suffixes for unsigned setp instructions.

Removed unused code.

Artem-B commented:
I still need to test it on live code, though we do not have much code that would end up using v4i8.

The generated PTX checked in llvm/test/CodeGen/NVPTX/i8x4-instructions.ll could use extra scrutiny.

Artem-B committed:
Merge branch 'llvm:main' into v4i8reg

ThomasRaoux commented on llvm/test/CodeGen/NVPTX/i8x4-instructions.ll:
```diff
@@ -0,0 +1,1248 @@
+; CHECK-NEXT:    .reg .b32 %r<3>;
+; CHECK-EMPTY:
+; CHECK-NEXT:  // %bb.0:
+; CHECK-NEXT:    ld.param.u32 %r1, [test_extract_0_param_0];
```
is it expected that the script checks for the hardcoded variable name?

Artem-B replied:
It's auto-generated, so it does check for everything that LLC spits out.

Artem-B replied:
In case you're referring to `test_extract_0_param_0`, it's the function parameter name, so yes, checking for it is expected.

ThomasRaoux commented:
I ran the patch on our triton kernels and I don't see any functional problems left.

ThomasRaoux reviewed (approved):
Looks like it required quite a lot of cases to be handled :( 
Thanks for doing this, it solves some of the problems triton had with latest LLVM. Changes look good to me.

Artem-B commented:
I see one suspicious failure in tensorflow tests. I suspect I've messed something up in v4i8 comparison.

Artem-B commented:
> I see one suspicious failure in tensorflow tests. I suspect I've messed something up in v4i8 comparison.

Yup, there is a problem:
```
Successfully custom legalized node
 ... replacing: t10: v4i8 = BUILD_VECTOR Constant:i16<-128>, Constant:i16<-128>, Constant:i16<-128>, Constant:i16<-128>
     with:      t17: v4i8 = bitcast Constant:i32<-128>
```

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2150,58 +2179,94 @@ NVPTXTargetLowering::LowerCONCAT_VECTORS(SDValue Op, SelectionDAG &DAG) const {
+  };
+  APInt Value;
+  if (Isv2x16VT(VT)) {
+    Value = GetOperand(Op, 0) | GetOperand(Op, 1).shl(16);
+  } else if (VT == MVT::v4i8) {
+    Value = GetOperand(Op, 0) | GetOperand(Op, 1).shl(8) |
+            GetOperand(Op, 2).shl(16) | GetOperand(Op, 3).shl(24);
```
This misses zext(32). Will fix shortly.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2150,58 +2179,94 @@ NVPTXTargetLowering::LowerCONCAT_VECTORS(SDValue Op, SelectionDAG &DAG) const {
+      Value = cast<ConstantSDNode>(Operand)->getAPIntValue();
+    else
+      llvm_unreachable("Unsupported type");
+    return Value.zext(32);
```
When we extract v4i8 values we must truncate them to 8 bits. They are carried in i16, so for negative values we end up with the unwanted extra 8 sign bits.

Artem-B committed:
Fixed calculation of constant v4i8 values.

Artem-B commented:
> > I see one suspicious failure in tensorflow tests. I suspect I've messed something up in v4i8 comparison.
> 
> Yup, there is a problem:
> 
> ```
> Successfully custom legalized node
>  ... replacing: t10: v4i8 = BUILD_VECTOR Constant:i16<-128>, Constant:i16<-128>, Constant:i16<-128>, Constant:i16<-128>
>      with:      t17: v4i8 = bitcast Constant:i32<-128>
> ```

Resolved by https://github.com/llvm/llvm-project/pull/67866/commits/9821e908e676d8eedbee7c07c90fb5aae4454f82

Artem-B committed:
Updated a test.

Artem-B committed:
Merge branch 'llvm:main' into v4i8reg

Artem-B commented:
Found another issue. We merge four independent byte loads with `align 1` into a 32-bit load, which fails at runtime on misaligned pointers. 

```
%t0 = type { [17 x i8] }

@shared_storage = linkonce_odr local_unnamed_addr addrspace(3) global %t0 undef, align 1

define <4 x i8> @in_v4i8(<4 x i8> %x, <4 x i8> %y) nounwind {
  %v = load <4 x i8>, ptr getelementptr inbounds (i8, ptr addrspacecast (ptr addrspace(3) @shared_storage to ptr), i64 9), align 1
  ret <4 x i8> %v
}
```

```
        mov.u64         %rd1, shared_storage;
        cvta.shared.u64         %rd2, %rd1;
        ld.u32  %r1, [%rd2+9];
        st.param.b32    [func_retval0+0], %r1;
        ret;
```

Artem-B committed:
Fixed unaligned load/store of v4i8

Artem-B committed:
Merge branch 'llvm:main' into v4i8reg

Artem-B committed:
clang-format

Artem-B commented:
clang-format failure on GitHub is weird -- it just silently exits with an error. 
I ran the same command locally and fixed one place it was not happy about.

The buildkite failure somewhere in RISC-V appears to be unrelated.

abadams commented:
I believe this may be causing failures for u/srem. See #69124

justinfargnoli commented on llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:
```diff
@@ -2110,6 +2214,29 @@ def : Pat<(seteq Int1Regs:$a, Int1Regs:$b),
+def: Pat<(setle (sext_inreg (trunc Int32Regs:$a), i8), (sext_inreg (trunc Int32Regs:$b), i8)),
+         (SETP_s32rr Int32Regs:$a, Int32Regs:$b, CmpLE)>;
+
+def: Pat<(setugt (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))),
+         (SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpHI)>;
+def: Pat<(setuge (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))),
+         (SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpHS)>;
+def: Pat<(setult (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))),
+         (SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpLO)>;
+def: Pat<(setule (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))),
+         (SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpLS)>;
+def: Pat<(seteq (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))),
+         (SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpEQ)>;
+def: Pat<(setne (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))),
+         (SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpNE)>;
```
Hi @Artem-B, I believe that these patterns are not correct.

Doing the comparison, without first ensuring that the value only uses the lower 8 bits, can lead to incorrect results. For example, when %a := Ox100 and %b := Ox1,
setugt (i16 (and (trunc Int32Regs:$a), 255)), (i16 (and (trunc Int32Regs:$b), 255))) produces a false predicate. Whereas SETP_u32rr Int32Regs:$a, Int32Regs:$b, CmpHI will produce a true predicate.

(I'm not super familiar with the pattern syntax through, so please correct me if I'm misunderstanding something.)
